<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Ibon Martínez-Arranz" /><link rel="canonical" href="https://github.com/imarranz/pygenmet/package/package/class.html" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Class - PyGenMet</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../../css/custom.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Class";
        var mkdocs_page_input_path = "package/package/class.md";
        var mkdocs_page_url = "/imarranz/pygenmet/package/package/class.html";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/python.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/bash.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../../index.html" class="icon icon-home"> PyGenMet
        </a>
        <div class="version">
          1.0
        </div><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../../index.html">PyGenMet</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">PyGenMet</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="class.html">Class</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="fitness.html">Fitness</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="operators.html">Operators</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="tools.html">Tools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Publications</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../publications/publications.html">Publications</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../publications/congress_conferences_seminars.html">Congress & Conferences</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../publications/metabolomics_subtypes.html">Metabolomics Subtypes</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">PyGenMet</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../index.html" class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>PyGenMet &raquo;</li>
      <li class="breadcrumb-item active">Class</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/imarranz/pygenmet/edit/master/docs/package/package/class.md">Edit on imarranz/pygenmet</a>
    </li>
  </ul>
    <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Breadcrumb Navigation">
        <a href="../introduction.html" class="btn btn-neutral float-left" title="Introduction"><span class="icon icon-circle-arrow-left" aria-hidden="true"></span> Previous</a>
        <a href="fitness.html" class="btn btn-neutral float-right" title="Fitness">Next <span class="icon icon-circle-arrow-right" aria-hidden="true"></span></a>
    </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="class_chromosome">CLASS <code>chromosome</code><a class="headerlink" href="#class_chromosome" title="Permanent link">#</a></h1>
<p>Import the <code>pygenmet</code> package</p>
<pre><code class="language-python">from pygenmet import *
</code></pre>
<p>With Pyhton a <code>class</code>can be created. A <code>class</code> provides a lot of functionalities. We can create a new <code>class</code> and then we can create a new type of object in Python. This new <code>class</code> allows new instances and these instances can also have <code>methods</code>.</p>
<p>The <code>class chromosome</code> has been created to optimization the algorithm. The <code>docstring</code> associated to the <code>chromosome</code> class is the following.</p>
<pre><code class="language-python">?chromosome
</code></pre>
<pre><code>The class chromosome has a copy of its genes, its fitness and 
the number of No NAFLD samples and NAFLD samples for print purposes.

Attributes
----------

Methods
-------

print()  
  Prints the chromosome

to_json()

Notes
-----
The representation of a chromosome is the following way

0110101101 ... 0110101101 || 0110110011 ... 0110110011     6     6           0.8945        
-------------------------    -------------------------   ----- -----         ------
            A                            B                 C     D              E

Description of the representation:
  A] Wild Type Genes
  B] Knock-Out Genes
  C] Number of Wild Type cases selected 
  D] Number of Knock-out cases selected
  E] The fitness value. The value of the adaptation for this chromosome

Examples
--------
&gt;&gt;&gt; CH = chromosome('01101011010110110011', 0.8945, 10, 10)
&gt;&gt;&gt; print("Genes: {}".format(CH.Genes))
Genes: 01101011010110110011        
&gt;&gt;&gt; print("WT Genes: {}".format(CH.WT_Genes))
WT Genes: 0110101101        
&gt;&gt;&gt; print("KO Genes: {}".format(CH.KO_Genes))
KO Genes: 0110110011        
&gt;&gt;&gt; print("Fitness: {}".format(CH.Fitness))
Fitness: 0.8945        
&gt;&gt;&gt; CH.print()
</code></pre>
<pre><code class="language-python">CH = chromosome('0110101101011000111101101011010110001111', 0.4125, 10, 30)
</code></pre>
<pre><code class="language-python">print(&quot;Genes: {}&quot;.format(CH.Genes))
</code></pre>
<pre><code>Genes: 0110101101011000111101101011010110001111
</code></pre>
<pre><code class="language-python">print(&quot;WT Genes: {}&quot;.format(CH.WT_Genes))
</code></pre>
<pre><code>WT Genes: 0110101101
</code></pre>
<pre><code class="language-python">print(&quot;KO Genes: {}&quot;.format(CH.KO_Genes))
</code></pre>
<pre><code>KO Genes: 011000111101101011010110001111
</code></pre>
<pre><code class="language-python">print(&quot;Fitness: {}&quot;.format(CH.Fitness))
</code></pre>
<pre><code>Fitness: 0.4125
</code></pre>
<h2 id="methods">Methods<a class="headerlink" href="#methods" title="Permanent link">#</a></h2>
<p>A Python method is like a Python function, except it is associated with object/classes. Methods in python are very similar to functions except for two major differences:</p>
<ul>
<li>The method is <strong>implicitly</strong> used for an object for which it is called.</li>
<li>The method is <strong>accessible</strong> to data that is contained within the class.</li>
</ul>
<p>The <em>chromosome</em> class has currently two methods:</p>
<ul>
<li>print()</li>
<li>to_json()</li>
</ul>
<h3 id="print">print()<a class="headerlink" href="#print" title="Permanent link">#</a></h3>
<p>The <em>print()</em> method allows to show the <code>chromosome</code> information on the screen or in a file.</p>
<p>The <em>print()</em> method has an argument, <em>N</em>, and with this argument we can customized the number of alleles to print. This method print first de <em>Wild Types</em> genes and second the <em>KO</em> genes separated by ||. The number of alleles equal to 1 for each gen is also printed. At last, the fitness function value is shown.</p>
<pre><code class="language-python">CH.print(N = 3) 
</code></pre>
<pre><code>011 ... 101 || 011 ... 111     6      18             0.4125
</code></pre>
<pre><code class="language-python">CH.print(N = 5) 
</code></pre>
<pre><code>01101 ... 01101 || 01100 ... 01111     6      18             0.4125
</code></pre>
<h3 id="to_json">to_json()<a class="headerlink" href="#to_json" title="Permanent link">#</a></h3>
<p><strong>JavaScript Object Notation</strong>, also known as <strong>JSON</strong>, is an open standard file format, that uses human-readable text to store and transmit data objects consisting of attribute–value pairs and array data types. It is also very easy for computers to parse and generate. The JSON is a very common data format.</p>
<pre><code class="language-python">CH.to_json(path_or_buf = 'out.json')
</code></pre>
<p>file <code>out.json</code> has the following structure:</p>
<pre><code class="language-json">{&quot;chromosome&quot;: [{&quot;Genes&quot;: &quot;0110101101011000111101101011010110001111&quot;, 
                 &quot;Fitness&quot;: 0.4125, 
                 &quot;WT_Genes&quot;: &quot;0110101101&quot;, 
                 &quot;KO_Genes&quot;: &quot;011000111101101011010110001111&quot;}]}
</code></pre>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p><b>PyGenMet: A Python Package for Genetic Algorithms</b> <br>Copyright &#169; 2024 <a href="https://github.com/imarranz">Ibon Martínez-Arranz</a><br></p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../introduction.html" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="fitness.html" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
