<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="author" content="Ibon Martínez-Arranz" /><link rel="canonical" href="https://github.com/imarranz/pygenmet/publications/metabolomics_subtypes.html" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Metabolomics Subtypes - PyGenMet</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../css/custom.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Metabolomics Subtypes";
        var mkdocs_page_input_path = "publications/metabolomics_subtypes.md";
        var mkdocs_page_url = "/imarranz/pygenmet/publications/metabolomics_subtypes.html";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/python.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/bash.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../index.html" class="icon icon-home"> PyGenMet
        </a>
        <div class="version">
          1.0
        </div><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../index.html">PyGenMet</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">PyGenMet</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../package/introduction.html">Introduction</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../package/package/class.html">Class</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../package/package/fitness.html">Fitness</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../package/package/operators.html">Operators</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../package/package/tools.html">Tools</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../package/package/examples.html">Examples</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Publications</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="publications.html">Publications</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="congress_conferences_seminars.html">Congress & Conferences</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="metabolomics_subtypes.html">Metabolomics Subtypes</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">PyGenMet</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html" class="icon icon-home" aria-label="Docs"></a> &raquo;</li>
          <li>Publications &raquo;</li>
      <li class="breadcrumb-item active">Metabolomics Subtypes</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/imarranz/pygenmet/edit/master/docs/publications/metabolomics_subtypes.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
    <div class="rst-breadcrumbs-buttons" role="navigation" aria-label="Breadcrumb Navigation">
        <a href="congress_conferences_seminars.html" class="btn btn-neutral float-left" title="Congress & Conferences"><span class="icon icon-circle-arrow-left" aria-hidden="true"></span> Previous</a>
    </div>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h3 id="metabolomics_subtypes">Metabolomics &amp; Subtypes<a class="headerlink" href="#metabolomics_subtypes" title="Permanent link">#</a></h3>
<p>Currently, there are three publications about subtypes of NASH with three mice models. The first publication was published in 2017.</p>
<h4 id="metabolomic_identification_of_subtypes_of_nonalcoholic_steatohepatitis">Metabolomic Identification of Subtypes of Nonalcoholic Steatohepatitis<a class="headerlink" href="#metabolomic_identification_of_subtypes_of_nonalcoholic_steatohepatitis" title="Permanent link">#</a></h4>
<h5 id="authors">Authors<a class="headerlink" href="#authors" title="Permanent link">#</a></h5>
<p>Cristina Alonso, David Fernández-Ramos, Marta Varela-Rey, <strong>Ibon Martínez-Arranz</strong>, Nicolás Navasa, Sebastiaan M Van Liempd, José L Lavín Trueba, Rebeca Mayo, Concetta P Ilisso, Virginia G de Juan, Marta Iruarrizaga-Lejarreta, Laura delaCruz-Villar, Itziar Mincholé, Aaron Robinson, Javier Crespo, Antonio Martín-Duce, Manuel Romero-Gómez, Holger Sann, Julian Platon, Jennifer Van Eyk, Patricia Aspichueta, Mazen Noureddin, Juan M Falcón-Pérez, Juan Anguita, Ana M Aransay, María Luz Martínez-Chantar, Shelly C Lu, José M Mato (2017). Metabolomic Identification of Subtypes of Nonalcoholic Steatohepatitis. <i>Gastroenterology</i>.</p>
<h5 id="abstract">Abstract<a class="headerlink" href="#abstract" title="Permanent link">#</a></h5>
<p>Nonalcoholic fatty liver disease (NAFLD) is a consequence of defects in diverse metabolic pathways that involve hepatic accumulation of triglycerides. Features of these aberrations might determine whether NAFLD progresses to nonalcoholic steatohepatitis (NASH). We investigated whether the diverse defects observed in patients with NAFLD are due to different NAFLD subtypes with specific serum metabolomic profiles, and whether these can distinguish patients with NASH from patients with simple steatosis.    </p>
<p>We collected liver and serum from methionine adenosyltransferase 1a knockout (MAT1A-KO) mice, which have chronically low level of hepatic S-adenosylmethionine (SAMe) and spontaneously develop steatohepatitis, as well as C57Bl/6 mice (controls); the metabolomes of all samples were determined. We also analyzed serum metabolomes of 535 patients with biopsy-proven NAFLD (353 with simple steatosis and 182 with NASH) and compared them with serum metabolomes of mice. MAT1A-KO mice were also given SAMe (30 mg/kg/day for 8 weeks); liver samples were collected and analyzed histologically for steatohepatitis.  </p>
<p>Livers of MAT1A-KO mice were characterized by high levels of triglycerides, diglycerides, fatty acids, ceramides, and oxidized fatty acids, as well as low levels of SAMe and downstream metabolites. There was a correlation between liver and serum metabolomes. We identified a serum metabolomic signature associated with MAT1A-KO mice that was also present in 49% of the patients; based on this signature, we identified 2 NAFLD subtypes. We identified specific panels of markers that could distinguish patients with NASH from patients with simple steatosis for each subtype of NAFLD. Administration of SAMe reduced features of steatohepatitis in MAT1A-KO mice.  </p>
<p>In an analysis of serum metabolomes of patients with NAFLD and MAT1A-KO mice with steatohepatitis, we identified 2 major subtypes of NAFLD and markers that differentiate steatosis from NASH in each subtype. These might be used to monitor disease progression and identify therapeutic targets for patients.</p>
<p><a href='https://www.gastrojournal.org/article/S0016-5085(17)30072-0/fulltext' style="color:white" class="btn btn-info">web</a>
<a href='https://www.gastrojournal.org/action/showPdf?pii=S0016-5085%2817%2930072-0' style="color:white" class="btn btn-info">pdf</a>
<a href='https://doi.org/10.1053/j.gastro.2017.01.015' style="color:white" class="btn btn-info">DOI</a>
<a href='https://pubmed.ncbi.nlm.nih.gov/28132890/' style="color:white" class="btn btn-info">Pubmed</a></p>
<h4 id="role_of_aramchol_in_steatohepatitis_and_fibrosis_in_mice">Role of aramchol in steatohepatitis and fibrosis in mice<a class="headerlink" href="#role_of_aramchol_in_steatohepatitis_and_fibrosis_in_mice" title="Permanent link">#</a></h4>
<h5 id="authors_1">Authors<a class="headerlink" href="#authors_1" title="Permanent link">#</a></h5>
<p>Marta Iruarrizaga-Lejarreta, Marta Varela-Rey, David Fernández-Ramos, <strong>Ibon Martínez-Arranz</strong>, Teresa C Delgado, Jorge Simon, Virginia Gutiérrez-de Juan, Laura delaCruz-Villar, Mikel Azkargorta, José L Lavin, Rebeca Mayo, Sebastiaan M Van Liempd, Igor Aurrekoetxea, Xabier Buqué, Donatella Delle Cave, Arantza Peña, Juan Rodríguez-Cuesta, Ana M Aransay, Felix Elortza, Juan M Falcón-Pérez, Patricia Aspichueta, Liat Hayardeny, Mazen Noureddin, Arun J Sanyal, Cristina Alonso, Juan Anguita, María Luz Martínez-Chantar, Shelly C Lu, José M Mato (2017). Role of aramchol in steatohepatitis and fibrosis in mice. <i>Hepatology Communications</i>.</p>
<h5 id="abstract_1">Abstract<a class="headerlink" href="#abstract_1" title="Permanent link">#</a></h5>
<p>Nonalcoholic steatohepatitis (NASH) is the advanced form of nonalcoholic fatty liver disease (NAFLD) that sets the stage for further liver damage. The mechanism for the progression of NASH involves multiple parallel hits, including oxidative stress, mitochondrial dysfunction, inflammation, and others. Manipulation of any of these pathways may be an approach to prevent NASH development and progression. Arachidyl-amido cholanoic acid (Aramchol) is presently in a phase IIb NASH study. The aim of the present study was to investigate Aramchol's mechanism of action and its effect on fibrosis using the methionine- and choline-deficient (MCD) diet model of NASH. We collected liver and serum from mice fed an MCD diet containing 0.1% methionine (0.1MCD) for 4 weeks; these mice developed steatohepatitis and fibrosis. We also collected liver and serum from mice receiving a control diet, and metabolomes and proteomes were determined for both groups. The 0.1MCD-fed mice were given Aramchol (5 mg/kg/day for the last 2 weeks), and liver samples were analyzed histologically. Aramchol administration reduced features of steatohepatitis and fibrosis in 0.1MCD-fed mice. Aramchol down-regulated stearoyl-coenyzme A desaturase 1, a key enzyme involved in triglyceride biosynthesis and the loss of which enhances fatty acid β-oxidation. Aramchol increased the flux through the transsulfuration pathway, leading to a rise in glutathione (GSH) and the GSH/oxidized GSH ratio, the main cellular antioxidant that maintains intracellular redox status. Comparison of the serum metabolomic pattern between 0.1MCD-fed mice and patients with NAFLD showed a substantial overlap. Conclusion: Aramchol treatment improved steatohepatitis and fibrosis by 1) decreasing stearoyl-coenyzme A desaturase 1 and 2) increasing the flux through the transsulfuration pathway maintaining cellular redox homeostasis. We also demonstrated that the 0.1MCD model resembles the metabolic phenotype observed in about 50% of patients with NAFLD, which supports the potential use of Aramchol in NASH treatment. (Hepatology Communications 2017).</p>
<p><a href='https://aasldpubs.onlinelibrary.wiley.com/doi/full/10.1002/hep4.1107' style="color:white" class="btn btn-info">web</a>
<a href='https://doi.org/10.1002/hep4.1107' style="color:white" class="btn btn-info">DOI</a>
<a href='https://pubmed.ncbi.nlm.nih.gov/29159325/' style="color:white" class="btn btn-info">Pubmed</a></p>
<h4 id="obeticholic_acid_modulates_serum_metabolites_and_gene_signatures_characteristic_of_human_nash_and_attenuates_inflammation_and_fibrosis_progression_in_ldlr--leiden_mice">Obeticholic Acid Modulates Serum Metabolites and Gene Signatures Characteristic of Human NASH and Attenuates Inflammation and Fibrosis Progression in Ldlr-/-.Leiden Mice<a class="headerlink" href="#obeticholic_acid_modulates_serum_metabolites_and_gene_signatures_characteristic_of_human_nash_and_attenuates_inflammation_and_fibrosis_progression_in_ldlr--leiden_mice" title="Permanent link">#</a></h4>
<h5 id="authors_2">Authors<a class="headerlink" href="#authors_2" title="Permanent link">#</a></h5>
<p>Martine C Morrison, Lars Verschuren, Kanita Salic, Joanne Verheij, Aswin Menke, Peter Y Wielinga, Marta Iruarrizaga-Lejarreta, Laurent Gole, Wei-Miao Yu, Scott Turner, Martien P M Caspers, <strong>Ibon Martínez-Arranz</strong>, Elsbet Pieterman, Reinout Stoop, Arianne van Koppen, Anita M van den Hoek, José M Mato, Roeland Hanemaaijer, Cristina Alonso, Robert Kleemann (2018). Obeticholic Acid Modulates Serum Metabolites and Gene Signatures Characteristic of Human NASH and Attenuates Inflammation and Fibrosis Progression in Ldlr-/-.Leiden Mice. <i>Hepatology Communications</i>, <i>2</i>(12), 1513–1532. <a href="https://doi.org/10.1002/hep4.1270">https://doi.org/10.1002/hep4.1270</a></p>
<h5 id="abstract_2">Abstract<a class="headerlink" href="#abstract_2" title="Permanent link">#</a></h5>
<p>Concerns have been raised about whether preclinical models sufficiently mimic molecular disease processes observed in nonalcoholic steatohepatitis (NASH) patients, bringing into question their translational value in studies of therapeutic interventions in the process of NASH/fibrosis. We investigated the representation of molecular disease patterns characteristic for human NASH in high-fat diet (HFD)-fed Ldlr-/-.Leiden mice and studied the effects of obeticholic acid (OCA) on these disease profiles. Multiplatform serum metabolomic profiles and genome-wide liver transcriptome from HFD-fed Ldlr-/-.Leiden mice were compared with those of NASH patients. Mice were profiled at the stage of mild (24 weeks HFD) and severe (34 weeks HFD) fibrosis, and after OCA intervention (24-34 weeks; 10 mg/kg/day). Effects of OCA were analyzed histologically, biochemically, by immunohistochemistry, using deuterated water technology (de novo collagen formation), and by its effect on the human-based transcriptomics and metabolomics signatures. The transcriptomics and metabolomics profile of Ldlr-/-.Leiden mice largely reflected the molecular signature of NASH patients. OCA modulated the expression of these molecular profiles and quenched specific proinflammatory-profibrotic pathways. OCA attenuated specific facets of cellular inflammation in liver (F4/80-positive cells) and reduced crown-like structures in adipose tissue. OCA reduced de novo collagen formation and attenuated further progression of liver fibrosis, but did not reduce fibrosis below the level before intervention. Conclusion: HFD-fed Ldlr-/-.Leiden mice recapitulate molecular transcriptomic and metabolomic profiles of NASH patients, and these signatures are modulated by OCA. Intervention with OCA in developing fibrosis reduces collagen deposition and de novo synthesis but does not resolve already manifest fibrosis in the period studied. These data show that human molecular signatures can be used to evaluate the translational character of preclinical models for NASH.</p>
<p><a href='https://aasldpubs.onlinelibrary.wiley.com/doi/full/10.1002/hep4.1270' style="color:white" class="btn btn-info">web</a>
<a href='https://doi.org/10.1002/hep4.1270' style="color:white" class="btn btn-info">DOI</a>
<a href='https://pubmed.ncbi.nlm.nih.gov/30556039/' style="color:white" class="btn btn-info">Pubmed</a></p>
              
            </div>
          </div><footer>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
      <p><b>PyGenMet: A Python Package for Genetic Algorithms</b> <br>Copyright &#169; 2024 <a href="https://github.com/imarranz">Ibon Martínez-Arranz</a><br></p>
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/imarranz/pygenmet" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="congress_conferences_seminars.html" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
